feat(api): Limited support for programmatically running PAPIv≥2.14

SyntaxColoring opened the pull request:
# Overview /  Changelog

These interfaces now support executing PAPIv≥2.14 protocols:

* The `opentrons_execute` CLI
* `opentrons.execute.execute()`
* `opentrons.execute.get_protocol_api()`

This works towards, but does not close, RSS-268.

Known limitations:

* *Simulating* PAPIv≥2.14 protocols will still raise a `NotImplementedError`.
* Executing or simulating *JSONv6+* protocols will still raise a `NotImplementedError`.
* It will not print the run log. (Ticketed as RSS-283.)

# Test Plan

## Regression testing

* [x] `opentrons_execute` with a PAPIv≤2.13 file
    * [x] Only standard labware
    * [x] Custom labware with `-L`
    * [x] Custom labware in the current working directory
* [x] `opentrons.execute.execute()` with a PAPIv≤2.13 file
    * [x] Only standard labware
    * [x] Custom labware with `custom_labware_paths=...`
    * [x] Custom labware in the Jupyter Notebook directory
* [x] `opentrons.execute.get_protocol_api()` with PAPIv≤2.13
    * [x] Only standard labware
    * [x] Custom labware with `extra_labware=...`
    * [x] Custom labware in the Jupyter Notebook directory

## New feature testing

* [x] `opentrons_execute` with a PAPIv≥2.14 file
    * [x] Only standard labware
    * [x] Custom labware with `-L`
    * [x] Custom labware in the current working directory
* [x] `opentrons.execute.execute()` with a PAPIv≥2.14 file
    * [x] Only standard labware
    * [x] Custom labware with `custom_labware_paths=...`
    * [x] Custom labware in the Jupyter Notebook directory
* [x] `opentrons.execute.get_protocol_api()` with PAPIv≥2.14
    * [x] Only standard labware
    * [x] Custom labware with `extra_labware=...`
    * [x] Custom labware in the Jupyter Notebook directory
* [x] Do at least one of these on an OT-3 

# Review requests

I hate several things about this.

* The `_LIVE_PROTOCOL_ENGINE_CONTEXTS` global.
* `_adapt_protocol_source()`.
* `create_protocol_engine_in_thread()`. This is technically not new—it was added in #12567—but now it's being used in production code for the first time.
* Letting `opentrons.execute` drift further apart from `opentrons.simulate`.

# Risk assessment

Medium. This is messy, highly-duplicated, difficult to automatically test, and public-facing.

Much of this code is sensitive to differences between `None` and `{}` for `bundled_labware`, `extra_labware`, and `bundled_data`.

# Next steps

We should try executing some Flex hardware testing scripts. From @ryanthecoder:

> `python3 -m hardware_testing.gravimetric --photometric --pipette 1000 --channels 96 --tip 200 --trials 5` this is how we call our scripts. the scripts themselves load a python protocol to fetch which type and location of labware but then just executes python directly talking to the protocol context

After that, we should continue working on RSS-268 until these interfaces are good enough to launch the Flex with.

SyntaxColoring committed:
Move async_context_manager_in_thread() to opentrons.utils.

This promotes this test helper to production code.

SyntaxColoring committed:
Minor improvements to async_context_manager_in_thread() and its tests.

* Fix a confusing assert with an incorrect message.
* Comment and refactor test_loop_lifetime() for clarity.
* Rename loop_queue to loop_mailbox to reflect what we're using it for.

SyntaxColoring committed:
Collect async_context_manager_in_thread() tests into a class.

We don't really do this often, but it seems nice for keeping the test function names short while still leaving room for testing the other functions in this file.

SyntaxColoring committed:
Move protocol_engine_in_thread() to opentrons.protocol_engine.

This promotes this test helper to production code.

SyntaxColoring committed:
Consistently take HardwareControlAPI interface.

SyntaxColoring committed:
Consistently take a full Config instead of trying to guess what the caller wants.

SyntaxColoring committed:
¯\_(ツ)_/¯

SyntaxColoring committed:
Add a hacky copy_file_like() helper.

SyntaxColoring committed:
Fix JSON protocols not being able to use any labware.

SyntaxColoring committed:
Fix Python protocols not including Jupyter labware by default.

SyntaxColoring committed:
Minor style fixups and comment clarifications.

SyntaxColoring committed:
Bundled data is supported, actually.

SyntaxColoring commented:
Note to self: Try to add tests for PAPIv2.14 in `test_execute.py`. Something is better than nothing.

SyntaxColoring committed:
Fix confusing error message.

SyntaxColoring committed:
Merge branch 'edge' into pe_based_execute

Merging for PR #13074.

SyntaxColoring commented on api/src/opentrons/execute.py:
```diff
@@ -5,60 +5,101 @@
 #: :py:meth:`get_protocol_api` will share
 
 
+# When a ProtocolContext is using a ProtocolEngine to control the robot, it requires some
+# additional long-lived resources besides _THREAD_MANAGED_HARDWARE. There's a background thread,
+# an asyncio event loop in that thread, and some ProtocolEngine-controlled background tasks in that
+# event loop.
+#
+# When we're executing a protocol file beginning-to-end, we can clean up those resources after it
+# completes. However, when someone gets a live ProtocolContext through get_protocol_api(), we have
+# no way of knowing when they're done with it. So, as a hack, we keep these resources open
+# indefinitely, letting them leak.
+#
+# We keep this at module scope so that the contained context managers aren't garbage-collected.
+# If they're garbage collected, they can close their resources prematurely.
+# https://stackoverflow.com/a/69155026/497934
+_LIVE_PROTOCOL_ENGINE_CONTEXTS = contextlib.ExitStack()
+
+
```
Reminder to self: Add an `atexit` handler for this, for consistency with `_THREAD_MANAGED_HW` and to eliminate unnecessary task leak warnings.

SyntaxColoring committed:
Don't attempt to pass bundled_labware in PAPIv2.14+.

sanni-t commented on api/src/opentrons/execute.py:
```diff
@@ -5,60 +5,101 @@
+
+
 # See Jira RCORE-535.
-_PYTHON_TOO_NEW_MESSAGE = (
-    "Python protocols with apiLevels higher than 2.13"
-    " cannot currently be executed with"
-    " the opentrons_execute command-line tool,"
-    " the opentrons.execute.execute() function,"
-    " or the opentrons.execute.get_protocol_api() function."
-    " Use a lower apiLevel"
-    " or use the Opentrons App instead."
-)
```
Maybe we can modify this message to talk about the limitations of the current implementation and log it as a warning instead. Just a thought

SyntaxColoring replied:
Discussed in-person, I think:

We'll raise fine-grained errors when, and only when, you try to do the specific things that the implementation doesn't yet support. I think we should keep them `NotImplementedError`s for now, instead of warnings.

SyntaxColoring committed:
Fix: Don't forget to add custom labware to PAPIv2.14+ live contexts.

SyntaxColoring committed:
Fix ^D hanging in a REPL after get_protocol_api("2.15").

sanni-t commented on api/src/opentrons/execute.py:
```diff
@@ -124,34 +171,39 @@ def get_protocol_api(
-    except protocol_api.ProtocolEngineCoreRequiredError as e:
-        raise NotImplementedError(_PYTHON_TOO_NEW_MESSAGE) from e  # See Jira RCORE-535.
+    else:
+        if bundled_labware is not None:
```
I don't know the context of bundled labware. Is it there for future implementation/completion or was it a part of some old version that we don't plan on supporting for newer versions? If it's the latter, we should update the error message to indicate that it's not supported anymore.

SyntaxColoring replied:
Yeah, I also don't know. Let's ask around and see if we can remove it.

SyntaxColoring committed:
Simplify the copy_file_like() hack by trying less.

SyntaxColoring committed:
Ugh.

SyntaxColoring committed:
Test fixup: Write encoding should match read encoding.

SyntaxColoring committed:
Test fixup: Use dummy binary data that's more obviously dummy & binary.

SyntaxColoring committed:
Test fixup: Rename test file to match source.

SyntaxColoring committed:
Test fixup: Remove unused param.

SyntaxColoring committed:
Merge branch 'edge' into pe_based_execute

Merging in case PR #13008 affects error handling.

SyntaxColoring committed:
Undo ":raises Exception:" docs addition.

This was just noise. I think it's okay to not document this explicitly, for now. We don't document other potential reasons why an execute would fail, and it's weird to treat this specially.

SyntaxColoring committed:
Fix errors not being reported.

SyntaxColoring committed:
Enable Protocol Engine in test parametrization.

SyntaxColoring committed:
Merge branch 'edge' into pe_based_execute

Merging for test and doc improvements in PR #13107.

SyntaxColoring commented on api/src/opentrons/execute.py:
```diff
@@ -401,24 +476,260 @@ def main() -> int:
+        super().__init__(errors)
+        self._error_occurrences = errors
+
+    def to_stderr_string(self) -> str:
+        """Return a string suitable as the stderr output of the `opentrons_execute` CLI.
+
+        This summarizes from the full error details.
+        """
+        # It's unclear what exactly we should extract here.
+        #
+        # First, do we print the first element, or the last, or all of them?
+        #
+        # Second, do we print the .detail? .errorCode? .errorInfo? .wrappedErrors?
+        # By contract, .detail seems like it would be insufficient, but experimentally,
+        # it includes a lot, like:
+        #
+        #     ProtocolEngineError [line 3]: Error 4000 GENERAL_ERROR (ProtocolEngineError):
+        #     UnexpectedProtocolError: Labware "fixture_12_trough" not found with version 1
+        #     in namespace "fixture".
+        return self._error_occurrences[0].detail
```
@TamarZanzouri @sfoster1 Is there a better thing for me to do here, now that stuff like #13008 is merged?

Here's what I'm trying to solve:

When we execute a protocol through Protocol Engine, we get any kind of fatal error—bugs, hardware errors, Protocol Engine validation errors, non-protocol-related Python mistakes, everything—through Protocol Engine's `state_summary.errors`. This is a list of `ErrorOccurrence`s.

From that, we need to decide:

1. How do we propagate that to callers of `opentrons.execute.execute()`? It needs to be an exception, but what do we put in it?

   I currently have the list of `ErrorOccurrence`s as its only member. In stack traces, Python will stringify that as the exception message. This seems...okay, I guess? It's structured and lossless, at least. But it's nonstandard. I guess I'd really prefer if it were a normal Python exception chain all the way down. 

2. What do we print to stderr for callers of the `opentrons_execute` CLI? Stringifying the `ErrorOccurrence`, or parts of it, seems plausible, but I don't know the details of how to do that properly.

SyntaxColoring committed:
Give up even harder on encodings, I guess.

SyntaxColoring committed:
Merge branch 'edge' into pe_based_execute

Fix merge conflicts in test_execute.py.

sanni-t commented on api/src/opentrons/execute.py:
```diff
@@ -401,24 +476,260 @@ def main() -> int:
+
+        # TODO(mm, 2023-06-30): This will home and drop tips at the end, which is not how
+        # things have historically behaved with PAPIv2.13 and older or JSONv5 and older.
+        result = await protocol_runner.run(protocol_source)
```
Oof, another problem with default/ forced tip dropping.

SyntaxColoring replied:
Yeah. the situation is actually not quite as bad as I thought it was, though. I thought we didn't have a way to clean up an engine without dropping tips. We do: we just have do do `ProtocolEngine.finish(drop_tips_and_home=False)`. The trouble is just at the `ProtocolRunner` level, which doesn't propagate that functionality up.

sanni-t commented on api/src/opentrons/execute.py:
```diff
@@ -401,24 +476,260 @@ def main() -> int:
+        robot_type=_get_robot_type(),
+        deck_type=DeckType(guess_deck_type_from_global_config()),
+        # We deliberately omit ignore_pause=True because, in the current implementation of
+        # opentrons.protocol_api.core.engine, that would incorrectly make
+        # ProtocolContext.is_simulating() return True.
```
Oo! I did not realize that's how we are checking for simulating engine. Looking at that code, maybe there's scope to reduce redundancy in some configs (virtual pipettes/ modules/ pause ignore) and pooling them all together into an `is_simulating` check (not for this PR). But I might be forgetting a reason why it was split into multiple flags.

SyntaxColoring replied:
Yeah, it seems like we should either split it *even further* and have a dedicated flag for something like `tell_python_protocol_it_is_simulating`, or merge them all. We're in a weird in-between right now.

sanni-t commented on api/src/opentrons/execute.py:
```diff
@@ -401,24 +476,260 @@ def main() -> int:
+        create_protocol_engine_in_thread(
+            hardware_api=hardware_api.wrapped(),
+            config=_get_protocol_engine_config(),
+            drop_tips_and_home_after=False,
```
So, am I understanding this correctly that live protocols like those over Jupyter will not drop tips and home at the end of the 'protocol session', but python/json protocol execution will?

SyntaxColoring replied:
Yep, exactly.

sanni-t reviewed (approved):
🙌 Awesome! I know it has a few hacks and still needs to be implemented fully, but this is an amazing start. This ticket is quite a complex puzzle and I think you've approached it quite carefully, covering a lot of edge cases. I might have missed a couple of things while reviewing too (re: it's a complex change), but unit tests and on-robot testing has been successful so I think this is merge worthy!

SyntaxColoring committed:
Don't test with emoji that can't be encoded by Windows' default encoding.

SyntaxColoring commented:
I'm seeing tests hang in a strange way at the closure of `pytest` on Windows, with Python 3.10, without `opentrons-hardware`. Investigating whether this is from the changes in this PR, or unrelated flakiness.

SyntaxColoring committed:
Perhaps.

SyntaxColoring committed:
Explain yoself.

SyntaxColoring commented:
I have a verbal 👍 from @TamarZanzouri on the above change to fix the shutdown hanging.
